File: source/includes/write/client-bulk-write.java
package org.example;

import com.mongodb.*;
import com.mongodb.client.model.*;
import com.mongodb.client.model.bulk.*;
import org.bson.Document;

import org.reactivestreams.Publisher;
import reactor.core.publisher.Mono;

import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoClients;

import java.util.*;

public class QuickStart {
    public static void main(String[] args) {
        // Replace the placeholder with your Atlas connection string
        String uri = "<connection string>";

        MongoClientSettings settings = MongoClientSettings.builder()
                .applyConnectionString(new ConnectionString(uri))
                .build();

        // Create a new client and connect to the server
        try (MongoClient mongoClient = MongoClients.create(settings)) {

            // start-insert-models
            ClientNamespacedInsertOneModel personToInsert = ClientNamespacedWriteModel
                    .insertOne(
                            new MongoNamespace("db", "people"),
                            new Document("name", "Julia Smith")
            );

            ClientNamespacedInsertOneModel thingToInsert = ClientNamespacedWriteModel
                    .insertOne(
                            new MongoNamespace("db", "things"),
                            new Document("object", "washing machine")
            );
            // end-insert-models

            // start-update-models
            ClientNamespacedUpdateOneModel personUpdate = ClientNamespacedWriteModel
                    .updateOne(
                            new MongoNamespace("db", "people"),
                            Filters.eq("name", "Freya Polk"),
                            Updates.inc("age", 1)
                    );

            ClientNamespacedUpdateManyModel thingUpdate = ClientNamespacedWriteModel
                    .updateMany(
                            new MongoNamespace("db", "things"),
                            Filters.eq("category", "electronic"),
                            Updates.set("manufacturer", "Premium Technologies")
                    );
            // end-update-models

            // start-replace-models
            ClientNamespacedReplaceOneModel personReplacement = ClientNamespacedWriteModel
                    .replaceOne(
                            new MongoNamespace("db", "people"),
                            Filters.eq("_id", 1),
                            new Document("name", "Frederic Hilbert")
                    );

            ClientNamespacedReplaceOneModel thingReplacement = ClientNamespacedWriteModel
                    .replaceOne(
                            new MongoNamespace("db", "things"),
                            Filters.eq("_id", 1),
                            new Document("object", "potato")
                    );
            // end-replace-models

            // start-perform
            MongoNamespace peopleNamespace = new MongoNamespace("db", "people");
            MongoNamespace thingsNamespace = new MongoNamespace("db", "things");

            List<ClientNamespacedWriteModel> bulkOperations = Arrays.asList(
                    ClientNamespacedWriteModel
                            .insertOne(
                                    peopleNamespace,
                                    new Document("name", "Corey Kopper")
                            ),
                    ClientNamespacedWriteModel
                            .replaceOne(
                                    thingsNamespace,
                                    Filters.eq("_id", 1),
                                    new Document("object", "potato")
                            )
            );
            
            Publisher<ClientBulkWriteResult> bulkWritePublisher = mongoClient
                    .bulkWrite(bulkOperations);

            ClientBulkWriteResult clientBulkResult = Mono
                    .from(bulkWritePublisher)
                    .block();
            
            System.out.println(clientBulkResult.toString());
            // end-perform

            // start-options
            MongoNamespace namespace = new MongoNamespace("db", "people");

            ClientBulkWriteOptions options = ClientBulkWriteOptions
                    .clientBulkWriteOptions()
                    .ordered(false);

            List<ClientNamespacedWriteModel> bulkOperations = Arrays.asList(
                    ClientNamespacedWriteModel.insertOne(
                            namespace,
                            new Document("_id", 1).append("name", "Rudra Suraj")
                    ),
                    // Causes a duplicate key error
                    ClientNamespacedWriteModel.insertOne(
                            namespace,
                            new Document("_id", 1).append("name", "Mario Bianchi")
                    ),
                    ClientNamespacedWriteModel.insertOne(
                            namespace,
                            new Document("name", "Wendy Zhang")
                    )
            );

            Publisher<ClientBulkWriteResult> bulkWritePublisher = mongoClient
                    .bulkWrite(bulkOperations, options);
            // end-options
        }
    }
}

File: current/source/includes/write/client-bulk-write.scala
package quickstart

import org.mongodb.scala._
import org.mongodb.scala.model.Filters._
import com.mongodb.client.model.bulk.{ClientBulkWriteOptions, ClientBulkWriteResult, ClientNamespacedWriteModel}
import org.mongodb.scala.model.Updates._

object Main {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string>")

    {
      // start-insert-models
      val personToInsert = ClientNamespacedWriteModel.insertOne(
        MongoNamespace("db", "people"),
        Document("name" -> "Julia Smith")
      )

      val thingToInsert = ClientNamespacedWriteModel.insertOne(
        MongoNamespace("db", "things"),
        Document("object" -> "washing machine")
      );
      // end-insert-models
    }
    {
      // start-update-models
      val personUpdate = ClientNamespacedWriteModel.updateOne(
        MongoNamespace("db", "people"),
        equal("name", "Freya Polk"),
        inc("age", 1)
      )

      val thingUpdate = ClientNamespacedWriteModel.updateMany(
        MongoNamespace("db", "things"),
        equal("category", "electronic"),
        set("manufacturer", "Premium Technologies")
      )
      // end-update-models
    }

    {
      // start-replace-models
      val personReplacement = ClientNamespacedWriteModel.replaceOne(
        MongoNamespace("db", "people"),
        equal("_id", 1),
        Document("name" -> "Frederic Hilbert")
      )

      val thingReplacement = ClientNamespacedWriteModel.replaceOne(
        MongoNamespace("db", "things"),
        equal("_id", 1),
        Document("object" -> "potato")
      )
      // end-replace-models
    }

    {
      // start-perform
      val peopleNamespace = MongoNamespace("db", "people")
      val thingsNamespace = MongoNamespace("db", "things")

      val writeModels = List(
        ClientNamespacedWriteModel.insertOne(
          peopleNamespace,
          Document("name" -> "Corey Kopper")
        ),
        ClientNamespacedWriteModel.replaceOne(
          thingsNamespace,
          equal("_id", 1),
          Document("object" -> "potato")
        )
      )

      val observable = mongoClient.bulkWrite(writeModels)

      observable.subscribe(
        (result: ClientBulkWriteResult) => println(result.toString),
        (error: Throwable) => println(s"Error: ${error.getMessage}"),
        () => println("Completed")
      )
      // end-perform
    }
    {
      // start-options
      val namespace = MongoNamespace("db", "people")

      val options = ClientBulkWriteOptions.clientBulkWriteOptions().ordered(false)

      val writeModels = List(
        ClientNamespacedWriteModel.insertOne(namespace, Document("_id" -> 1, "name" -> "Rudra Suraj")),
        // Causes a duplicate key error
        ClientNamespacedWriteModel.insertOne(namespace, Document("_id" -> 1, "name" -> "Mario Bianchi")),
        ClientNamespacedWriteModel.insertOne(namespace, Document("name" -> "Wendy Zhang"))
      )

      val observable = mongoClient.bulkWrite(writeModels, options)
      // end-options
    }

    Thread.sleep(1000)
    mongoClient.close()

  }

}
